### PR TITLE
Client Suchfunktion

### PR DESCRIPTION
**Pull request by casaper**
# [Story in Trello](https://trello.com/c/Uz0BSVr5/42-suchfunktion)

## ToDo:

- [x] Find way to acceptance (_system_) test jquery.ui.autocomplete
- [x] integration tests on JSON endpoint

## Questions

- Are the autofocus and close autosubmit in the JS a help or a dept?

### Done for approval?

* [x] Acceptance criteria are met
* [x] Code quality checks are green
* [x] Readme updated if needed
* [x] Story under test
* [x] Mobile works
* [ ] Seeds created
* [x] Code is reviewed

### Done after the merge?

* [ ] Deployed to staging after the merge
* [ ] Tested manually on staging

### CODE REVIEW

File: app/models/contact.rb [hunk 1]
```diff
@@ -66,6 +64,14 @@ def validate_last_name?
     !profile?
   end
 
+  def full_name_changed?
+    first_name_changed? || last_name_changed?
```
Comment (chrysalag) [inline app/models/contact.rb]: `saved_change_to_first_name? || saved_change_to_last_name?`

File: app/views/clients/search.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (chrysalag) [inline app/views/clients/search.json.jbuilder]: NIT (RuboCop) picking: Remove empty line

File: app/models/contact.rb (hunk 1 above)
Comment (casaper) [reply app/models/contact.rb]: if my callback is before_save?

Comment (chrysalag) [reply app/models/contact.rb]: Then maybe `will_save_change_to_attribute?`

Comment (casaper) [reply app/models/contact.rb]: Ah, now I realize, it is depricated, yea. But my reaction shows that the Rails team did a extremly bad job with explicity and naming there.

**Comment by casaper** [comment]
@nuritnt middle or end of the word is the prequisit for fullfilling a first or lastname search in that manner. Everything else would be way more complicated or end up beeing a start of lastname or end of first name search after all

File: test/models/contact_test.rb [hunk 3]
```diff
@@ -41,4 +41,23 @@ def setup
     refute @contact.valid?
     assert_equal @contact.errors.keys, [:last_name]
   end
+
+  test 'update_first_or_lastname_changes_full_name_atribute' do
```
Comment (chrysalag) [inline test/models/contact_test.rb]: attribute

File: test/system/client_searches_test.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    assert page.has_text? @clients.last.first.contact.full_name
+    assert page.has_text? @clients.last.last.contact.full_name
+  end
+
+  test 'enter_search_text_brings_suggestions' do
+    fill_autocomplete 'q[contact_full_name_cont]', with: 'aaa', items_expected: 2,
+      check_items: [@clients.first[0].contact.full_name, @clients.first[1].contact.full_name]
+  end
+
+  test 'suggettions search triggers the search correctly' do
```
Comment (chrysalag) [inline test/system/client_searches_test.rb]: suggestions

File: app/assets/javascripts/search.es6 [hunk 5]
```diff
@@ -0,0 +1,25 @@
+$(document).on('turbolinks:load', () => {
+  $("input[data-autocomplete]").each((_index, input) => {
+    $(input).autocomplete({
+      source: $(input).data('autocomplete'),
+      close: handleCloseAutosuggest,
+      focus: handleAutosuggestFocus
+    });
+
+    // Select all text in input, for easyer deleting of all content if desired
```
Comment (chrysalag) [inline app/assets/javascripts/search.es6]: easier

**Pull request merged**